<a href="https://colab.research.google.com/github/parimalakettymuthu/MachineLearning-Projects/blob/main/Multiclass_classification_Irirs_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from sklearn import datasets
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split

In [3]:
iris_data = datasets.load_iris()  #Loaded the iris dataset  

In [5]:
iris_data.DESCR

'.. _iris_dataset:\n\nIris plants dataset\n--------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n                \n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)\n    ============== ==== =

In [7]:
iris_data.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [9]:
iris_data.data[0]

array([5.1, 3.5, 1.4, 0.2])

In [10]:
iris_data.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [11]:
iris_data.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [12]:
iris_data.data.shape

(150, 4)

In [13]:
X_var = iris_data.data    #Defined the X & labels (y)
y_var = iris_data.target

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X_var, y_var, stratify = y_var, random_state=42)

In [20]:
## DecisionTree classifier 
from sklearn.tree import DecisionTreeClassifier
mc_tree = DecisionTreeClassifier(max_depth=2).fit(X_train, y_train)
mc_tree

DecisionTreeClassifier(max_depth=2)

In [32]:
mc_tree.max_features_

4

In [36]:
mc_tree.feature_importances_

array([0., 0., 0., 1.])

In [37]:
mc_tree.get_params

<bound method BaseEstimator.get_params of DecisionTreeClassifier(max_depth=2)>

In [34]:
#Run the prediction on test dataset
mc_tree_pred = mc_tree.predict(X_test)
mc_tree_pred   #Predicted label

array([0, 1, 1, 1, 0, 1, 1, 2, 2, 2, 2, 2, 1, 1, 0, 0, 0, 1, 0, 1, 2, 1,
       2, 1, 2, 1, 0, 2, 0, 2, 2, 2, 0, 0, 0, 0, 2, 1])

In [21]:
#Ground truth label
y_test   

array([0, 1, 1, 1, 0, 1, 2, 2, 2, 2, 2, 2, 1, 1, 0, 0, 0, 1, 0, 1, 2, 1,
       2, 1, 2, 1, 0, 2, 0, 1, 2, 2, 0, 0, 0, 0, 2, 1])

In [23]:
mc_tree_cm = confusion_matrix(y_test, mc_tree_pred)
print("Decision Tree Confusion matrix: ", mc_tree_cm)

Decision Tree Confusion matrix:  [[12  0  0]
 [ 0 12  1]
 [ 0  1 12]]


In [61]:
#Decision Tree Accuracy
print("Decision tree accuracy: ", mc_tree.score(X_test, y_test))

Decision tree accuracy:  0.9473684210526315


In [28]:
## Training using svm
from sklearn.svm import SVC
mc_svm = SVC(kernel='linear', C=1).fit(X_train, y_train)
mc_svm.coef_

array([[-0.16460697,  0.53667478, -0.82497057, -0.62238253],
       [-0.15447432,  0.15924442, -0.4652579 , -0.22984037],
       [ 0.58803505,  0.27328846, -1.90524516, -1.79800552]])

In [29]:
mc_svm.intercept_

array([1.65737446, 2.1586797 , 7.87379436])

In [31]:
mc_svm.get_params()

{'C': 1,
 'break_ties': False,
 'cache_size': 200,
 'class_weight': None,
 'coef0': 0.0,
 'decision_function_shape': 'ovr',
 'degree': 3,
 'gamma': 'scale',
 'kernel': 'linear',
 'max_iter': -1,
 'probability': False,
 'random_state': None,
 'shrinking': True,
 'tol': 0.001,
 'verbose': False}

In [39]:
mc_svm_pred = mc_tree.predict(X_test)   #Predicting the labels using X_test dataset
mc_svm_pred

array([0, 1, 1, 1, 0, 1, 1, 2, 2, 2, 2, 2, 1, 1, 0, 0, 0, 1, 0, 1, 2, 1,
       2, 1, 2, 1, 0, 2, 0, 2, 2, 2, 0, 0, 0, 0, 2, 1])

In [40]:
mc_svm_cm = confusion_matrix(y_test, mc_svm_pred)  #Confusion matrix for svm
print("Confusion matrix of SVM: ", mc_svm_cm)

Confusion matrix of SVM:  [[12  0  0]
 [ 0 12  1]
 [ 0  1 12]]


In [60]:
#SVM Accuracy 
print("SVM Accuracy: ", mc_svm.score(X_test, y_test))

SVM Accuracy:  1.0


In [45]:
##Implementing KNN classifier 
from sklearn.neighbors import KNeighborsClassifier
mc_knn = KNeighborsClassifier(n_neighbors = 7).fit(X_train, y_train)
mc_knn.kneighbors

<bound method KNeighborsMixin.kneighbors of KNeighborsClassifier(n_neighbors=7)>

In [46]:
mc_knn.get_params

<bound method BaseEstimator.get_params of KNeighborsClassifier(n_neighbors=7)>

In [47]:
mc_knn.n_features_in_

4

In [48]:
mc_knn_pred = mc_knn.predict(X_test)

In [51]:
print("Accuracy: ", mc_knn.score(X_test, y_test))

Accuracy:  0.9473684210526315


In [52]:
mc_knn_cm = confusion_matrix(y_test, mc_knn_pred)
print("KNeighbors classifier confusion matrix: ", mc_knn_cm)

KNeighbors classifier confusion matrix:  [[12  0  0]
 [ 0 13  0]
 [ 0  2 11]]


In [54]:
#Implementing Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB
mc_gnb = GaussianNB().fit(X_train, y_train)   #Created GNB instance and train the model using train dataset
mc_gnb.n_features_in_

4

In [56]:
mc_gnb_pred = mc_gnb.predict(X_test)  #Predict the X_test dataset

In [57]:
mc_gnb_pred

array([0, 1, 1, 1, 0, 1, 1, 2, 2, 2, 1, 2, 1, 1, 0, 0, 0, 1, 0, 1, 2, 1,
       2, 1, 2, 1, 0, 2, 0, 2, 2, 2, 0, 0, 0, 0, 2, 1])

In [58]:
mc_gnb_cm = confusion_matrix(y_test, mc_gnb_pred)  #Confusion matrix
print("Confusion matrix of GNB: ", mc_gnb_cm)

Confusion matrix of GNB:  [[12  0  0]
 [ 0 12  1]
 [ 0  2 11]]


In [59]:
#Accuracy
print("GNB accuracy: ", mc_gnb.score(X_test, y_test))

GNB accuracy:  0.9210526315789473


There's no much difference between the confusion matrix generated between these algorithms; Accuracy has been varied: 
1. Decision Tree accuracy: 0.9473684210526315
2. SVM accuracy: 1.0
3. KNeighbors classifier accuracy: 0.9473684210526315
4. Gaussian Naive Bayes accuracy: 0.9210526315789473

SVM have given a perfect score of 1 accuracy; 
Decision Tree & KNeighbors stands second with .94 accuracy; More data can be used to determine the best algorithm based on accuracy and f1 score
